spatial/r3/mat.go

```diff
@@ -197,3 +197,52 @@ func (m *Mat) Det() float64 {
 	detc := m.At(1, 0)*m.At(2, 1) - m.At(1, 1)*m.At(2, 0)
 	return a*deta - b*detb + c*detc
 }
+
+// Hessian sets the receiver to the hessian matrix of a spatial scalar field
+// evaluated at point p. Additional points evaluated to calculate hessian
+// are spaced at most tol away from p.
+func (m *Mat) Hessian(p, tol Vec, f func(Vec) float64) {
+	dx := Vec{X: tol.X}
+	dy := Vec{Y: tol.Y}
+	dz := Vec{Z: tol.Z}
+	fp := f(p)
+	fxp := f(Add(p, dx))
+	fxm := f(Sub(p, dx))
+	fxx := (fxp - 2*fp + fxm) / (tol.X * tol.X)
+
+	fyp := f(Add(p, dy))
+	fym := f(Sub(p, dy))
+	fyy := (fyp - 2*fp + fym) / (tol.Y * tol.Y)
+
+	var aux Vec
+	aux = Add(dx, dy)
+	fxyp := f(Add(p, aux))
+	fxym := f(Sub(p, aux))
+	fxy := (fxyp - fxp - fyp + 2*fp - fxm - fym + fxym) / (2 * tol.X * tol.Y)
```

Comment:
```suggestion
	fxy := (fxyp - fxp - fyp + 2*fp - fxm - fym + fxym) * 0.5 * iy2
```

Reply:
This may be invalid. The cross partial derivative of x and y requires being divided by both the steps in the x and y direction, which may not be equal. From https://en.wikipedia.org/wiki/Finite_difference#Multivariate_finite_differences